Add a NeuroVault fetcher

bcipolli opened the pull request:
This is an enhancement for #640. Not quite done, but far enough along to talk about the API I've implemented.

To do:
* [x] Implement an API for downloading selected images from http://neurovault.org/
* [x] Create dataset documentation
* [x] Create a basic example for downloading a subset of files.
* [x] Create a more complex ICA-based example.
* [x] Create tests
* [x] Remove dependency on `requests` - use `urllib`
* [x] Read from disk when offline.
* [ ] Detect outdated cached collection or image metadata or data.

Neurovault contains images in collections. Collections can be filtered via a few querystring arguments. Images to download and return should be limited by filters on collection and image metadata.

To do this, I've included `collection_filters` and `image_filters` arguments.
* If `isinstance(collection_filters, dict)`, the filters are used as querystring argument filters.
* If `isinstance(*_filters, collections.Iterable)` (i.e. a list), the filters are assumed to be functions that receive a collection or image metadata, and returns `True` if the collection's images should be queried, or if the image should be downloaded.

The function returns a `Bunch` of three:
* `collections`: a dict (key: collection_id) of collection metadata for matched collections.
* `images`: a list of image metadata (including local file paths) for matched & downloaded images.
* `func_files`: a list of paths to downloaded files. Could be removed, but I kept it as many other functions return a list of paths to images.

The example I've begun with is a `stop-go` task meta-analysis, ported from:
https://github.com/NeuroVault/neurovault_analysis/blob/master/Stop%20vs.%20Go%20metaanalysis.ipynb

Note that I avoided trying to convert things to a `pandas` dataframe. The previous examples downloaded all images and filtered using `pandas`; however, all `nilearn` dataset fetchers filter before returning to the user. It was simple enough to allow lambda functions as filters to make very powerful pre-fetch filtering available to the user, so I went that route.

Looking for feedback on the overall API design to start. @chrisfilo and anyone else around!

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1279,229 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+            url = base_url
+        return url
+
+    def _get_nv_json(url, local_file=None, refresh=False):
```
Why don't you use the file fetching functions of datasets here? I guess that we will have a problem of update with neurovault: a given request may return different results if the neurovault base has been updated. We should maybe use a flag to work online / offline.

bcipolli replied:
Yep, should definitely try to migrate this. If it requires tweaks to the fetching functions, will submit those as a separate PR.

AlexandreAbraham commented:
> Not quite done, but far enough along to talk about the API I've implemented.

I think that the macanism of filters is too complicated for most of our users. We can keep it for advanced users but at some point we should provide simpler pre-made filters. For example, I think that a user should be able to specify the id of a collection or a list of ids (I know that it's possible through collection_filter). But we'll see that later.

> The example I've begun with is a stop-go task meta-analysis

Reproducing an example provided by neurovault is indeed the way to go.

> Note that I avoided trying to convert things to a pandas dataframe. 

So far, we have avoided this dependency but we may have to include it at some point. Is there a real benefit to use pandas here?

The PR so far looks like really good work to me ;).

bcipolli commented:
>We should maybe use a flag to work online / offline.

Regardless of online or offline, there is an updating problem. When online, the code assumes that a cached copy of a particular collection or image metadata, or image data is correct / up-to-date. I have the `refresh` flag for controlling that behavior. A more automated mechanism is possible.

Separate from that, the code needs to work in offline mode; currently it does not. I have added that to the WIP item list.

>I think that the macanism of filters is too complicated for most of our users. We can keep it for advanced users but at some point we should provide simpler pre-made filters. For example, I think that a user should be able to specify the id of a collection or a list of ids (I know that it's possible through collection_filter). But we'll see that later.

Right; glad to add more simple filters. It's very easy to map from the simple filters to the more complex ones. The complex filters were needed to mirror the logic of the example I ported, but I could also change it to use specific collections and images (and tell users they were determined by searches on the website).

>So far, we have avoided this dependency but we may have to include it at some point. Is there a real benefit to use pandas here?

In the previous examples, the benefit was using Pandas' filtering mechanisms, which tend to lead to simpler code. But I just don't see it as an option here, unless we limit users to very simple filters (which I think would be a shame).

>The PR so far looks like really good work to me ;).

:D

chrisfilo commented:
Looks good to me so far! Let me know if you have any questions about the API (we can make improvements!)

bcipolli commented:
Thanks @chrisfilo! One thing that might help is if the API endpoint could return headers indicating the response size; `nilearn` doesn't like not getting a `Content-Length` header, I believe.

To reduce the number of requests made, it could also be helpful to be able to get image metadata for a specified set of collection IDs, rather than one at a time.

chrisfilo commented:
On Tue, Nov 10, 2015 at 9:02 AM, Ben Cipollini <notifications@github.com>
wrote:

> Thanks @chrisfilo <https://github.com/chrisfilo>! One thing that might
> help is if the API endpoint could return headers indicating the response
> size; nilearn doesn't like not getting a Content-Length header, I believe.
>
That's for a single file download or part of JSON?


> To reduce the number of requests made, it could also be helpful to be able
> to get image metadata for a specified set of collection IDs, rather than
> one at a time.
>
Hmm - we can work on adding this.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/832#issuecomment-155487633>.
>

bcipolli commented:
@AlexandreAbraham in order to remove the `requests` dependency, I needed to make changes to `_fetch_files` (#838 and #839; add overwrite option, modify output). Not sure this is a net positive; the code was simpler all around when using `requests`. Glad to get feedback!

bcipolli commented:
>That's for a single file download or part of JSON?
JSON. Downloading actual images works perfectly.

AlexandreAbraham commented:
> nilearn doesn't like not getting a Content-Length header, I believe.

This case should be handled properly.

> the code was simpler all around when using requests

I am not against adding requests as a dependency but you should ask to the big boss :P.

bcipolli commented:
@chrisfilo Another issue I'm having with the NeuroVault API is that collection ID returned seems unordered. This means that, when I'm offline, I can't order collections as I get them when I'm online. This causes problems.

Is it possible to get collection IDs returned in order?

Here's what I see for collection IDs:

```
Downloading data from http://neurovault.org/api/collections  ...
 [10, 862, 38, 55, 72, 75, 81, 82, 84, 554, 14, 71, 816, 723, 440, 556, 35, 67, 17, 66, 471, 92, 751, 4, 923, 496, 70, 921, 79, 93, 863, 89, 609, 542, 497, 784, 57, 109, 102, 114, 117, 119, 610, 103, 465, 612, 118, 115, 443, 494, 215, 882, 787, 795, 790, 105, 140, 796, 800, 794, 804, 865, 506, 495, 33, 308, 473, 500, 927, 510, 907, 809, 588, 625, 508, 190, 631, 572, 812, 573, 628, 231, 574, 295, 298, 292, 276, 294, 305, 408, 285, 287, 885, 319, 821, 356, 409, 323, 324, 329]

Downloading data from http://neurovault.org/api/collections/?offset=100 ...
[586, 331, 320, 340, 367, 335, 336, 886, 585, 337, 344, 360, 365, 376, 377, 929, 379, 814, 46, 391, 555, 822, 559, 576, 931, 887, 581, 486, 589, 594, 297, 597, 592, 598, 932, 934, 591, 903, 942, 601, 599, 596, 562, 448, 825, 889, 43, 398, 912, 551, 933, 563, 59, 187, 179, 94, 552, 451, 460, 826, 890, 648, 913, 203, 7, 418, 419, 550, 307, 650, 646, 461, 34, 640, 227, 641, 639, 915, 684, 553, 687, 655, 545, 681, 683, 315, 686, 893, 3, 527, 434, 526, 806, 515, 697, 700, 708, 709, 831, 519]
```

chrisfilo commented:
Hmm. I can add this, but it will slow down the API (not sure how much). Why
is the order necessary?

On Sun, Nov 15, 2015 at 8:27 PM, Ben Cipollini <notifications@github.com>
wrote:

> @chrisfilo <https://github.com/chrisfilo> Another issue I'm having with
> the NeuroVault API is that collection ID returned seems unordered. This
> means that, when I'm offline, I can't order collections as I get them when
> I'm online. This causes problems.
>
> Is it possible to get collection IDs returned in order?
>
> Here's what I see for collection IDs:
>
> Downloading data from http://neurovault.org/api/collections  ...
>  [10, 862, 38, 55, 72, 75, 81, 82, 84, 554, 14, 71, 816, 723, 440, 556, 35, 67, 17, 66, 471, 92, 751, 4, 923, 496, 70, 921, 79, 93, 863, 89, 609, 542, 497, 784, 57, 109, 102, 114, 117, 119, 610, 103, 465, 612, 118, 115, 443, 494, 215, 882, 787, 795, 790, 105, 140, 796, 800, 794, 804, 865, 506, 495, 33, 308, 473, 500, 927, 510, 907, 809, 588, 625, 508, 190, 631, 572, 812, 573, 628, 231, 574, 295, 298, 292, 276, 294, 305, 408, 285, 287, 885, 319, 821, 356, 409, 323, 324, 329]
>
> Downloading data from http://neurovault.org/api/collections/?offset=100 ...
> [586, 331, 320, 340, 367, 335, 336, 886, 585, 337, 344, 360, 365, 376, 377, 929, 379, 814, 46, 391, 555, 822, 559, 576, 931, 887, 581, 486, 589, 594, 297, 597, 592, 598, 932, 934, 591, 903, 942, 601, 599, 596, 562, 448, 825, 889, 43, 398, 912, 551, 933, 563, 59, 187, 179, 94, 552, 451, 460, 826, 890, 648, 913, 203, 7, 418, 419, 550, 307, 650, 646, 461, 34, 640, 227, 641, 639, 915, 684, 553, 687, 655, 545, 681, 683, 315, 686, 893, 3, 527, 434, 526, 806, 515, 697, 700, 708, 709, 831, 519]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/832#issuecomment-156911793>.
>

bcipolli commented:
I have a parameter to limit the number of images returned. If I iterate through collections differently online and offline, the image set will be different between the two.

chrisfilo commented:
Could you give me some more context? What do you mean by iterating over
collections offline? Is this for synchronization? Why not download all
collection metadata and then perform updates/deletions?

On Sun, Nov 15, 2015 at 8:44 PM, Ben Cipollini <notifications@github.com>
wrote:

> I have a parameter to limit the number of images returned. If I iterate
> through collections differently online and offline, the image set will be
> different between the two.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/832#issuecomment-156914380>.
>

bcipolli commented:
>What do you mean by iterating over collections offline?

The case where a user successfully accesses neurovault online. We cache the data to disk so that if they go offline, they have the data available to run analyses.

> Is this for synchronization?

Nope. Just to run some analysis. I don't actually store the giant json file--I decimate it on a per-collection basis. The directory structure is: `[collection-id]/[image-id]/image_files`, with collection metadata stored inside the appropriate collection directory. Seemed cleaner to store the collection metadata that way, than in paged chunks.

Works great, except I lose the order of collections doing that. I expected they'd be returned in a meaningful order. I was wrong lol


If the user limits the number of images returned (which I have a parameter to do), then to get the same image set for the analysis in the online & offline case, I have to iterate through the collection metadata in the same order as it is returned when I'm online., because I want to allow users to limit the number of images they get back.

chrisfilo commented:
Thanks this is much clearer.

We use limits and paging to manage transfer of a lot of metadata. I'm not
sure how useful this would be for anything else. For example I don't quite
understand why anyone would like to do analysis on 30 first collections in
NV.

A more realistic use cases would include:
- limiting collection to only those that have paper DOIs (because those are
supposedly more trustworthy)
- excluding single subejct and metanalysis images
- excluding thresholded  and non MNI images

Don't get me wrong - I'm happy to add the ordering functionality, but I'm
not sure how useful this would be. Why would anyone analyze the first 30
oldest collections? Or 50 most recent ones?

On Sun, Nov 15, 2015 at 8:58 PM, Ben Cipollini <notifications@github.com>
wrote:

> What do you mean by iterating over collections offline?
>
> The case where a user successfully accesses neurovault online. We cache
> the data to disk so that if they go offline, they have the data available
> to run analyses.
>
> Is this for synchronization?
>
> Nope. Just to run some analysis. I don't actually store the giant json
> file--I decimate it on a per-collection basis. The directory structure is:
> [collection-id]/[image-id]/image_files, with collection metadata stored
> inside the appropriate collection directory. Seemed cleaner to store the
> collection metadata that way, than in paged chunks.
>
> Works great, except I lose the order of collections doing that. I expected
> they'd be returned in a meaningful order. I was wrong lol
>
> If the user limits the number of images returned (which I have a parameter
> to do), then to get the same image set for the analysis in the online &
> offline case, I have to iterate through the collection metadata in the same
> order as it is returned when I'm online., because I want to allow users to
> limit the number of images they get back.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/832#issuecomment-156915824>.
>

bcipolli commented:
>I'm not sure how useful this would be. Why would anyone analyze the first 30 oldest collections? Or 50 most recent ones?

The limit is on the # of images returned. I use it as data download is slow, so it lets me develop analyses on a subset of the data, then to run the full analysis when it's done. It also gives some safety when filtering collections & images, so I don't download a huge dataset unexpectedly.

All the filtering you mention is possible through the current interface.

Do you think the `max_images` parameter should simply be eliminated, then?

chrisfilo commented:
I guess the max_images would be useful for tests. Do we need order for
those though?

On Sun, Nov 15, 2015 at 9:22 PM, Ben Cipollini <notifications@github.com>
wrote:

> I'm not sure how useful this would be. Why would anyone analyze the first
> 30 oldest collections? Or 50 most recent ones?
>
> The limit is on the # of images returned. I use it as data download is
> slow, so it lets me develop analyses on a subset of the data, then to run
> the full analysis when it's done. It also gives some safety when filtering
> collections & images, so I don't download a huge dataset unexpectedly.
>
> All the filtering you mention is possible through the current interface.
>
> Do you think the max_images parameter should simply be eliminated, then?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/832#issuecomment-156917991>.
>

bcipolli commented:
Yeah, I'll think more about the different use-cases and let you know. Thanks for chatting about it with me!

GaelVaroquaux commented on examples/manipulating_visualizing/plot_neurovault_meta_analysis.py:
```diff
@@ -0,0 +1,104 @@
+from nilearn.plotting import plot_glass_brain
+for im in images:
+    plot_glass_brain(im['local_path'], title='%04d' % im['id'])
+plt.show()
```
nitpick, but I think that I would rather avoid having calls to 'show' in the middle.

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1279,269 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    data_dir = _get_dataset_dir('neurovault',
+                                data_dir=data_dir)
+
+    def _build_nv_url(base_url, filts=None):
```
I find that it is generally better to move these function outside of the enclosing scope, so that they can tested, debugged, and maybe one day reused more easily.

GaelVaroquaux commented:
I am not sure why tests are failing. But I did check on master, and they do not fail.

GaelVaroquaux commented:
By the way: I few nitpicks, but in general, this PR looks really really great. Thanks

lesteve commented:
> I am not sure why tests are failing. 

The Travis one fails for the Python 2.6 distrib. I am guessing the .header property doesn't exist for nibabel 1.1.0.

bcipolli commented:
>By the way: I few nitpicks, but in general, this PR looks really really great. Thanks

Thanks @GaelVaroquaux! I'll fix them up tonight. I also have the basics done of porting the ICA example. 

I will update here when I've updated the code per your comments, fixed the tests, and have downloaded all the data and verified my ICA results look like the published ones.

bcipolli commented:
By the way, y'all have been on my mind with all that's been happening in Paris. Hope all your family, friends, and colleagues are safe. :+1:

bcipolli commented:
OK, I believe this ready for consideration to merge. I've implemented both examples and am satisfied with the code. 

Currently, if a collection is updated in NeuroVault, it will not be updated locally. The only way to get around this is to pass the `overwrite=True` flag, which will re-download all metadata and images. If efficient local updating is desired, I suggest opening a new issue for it; I think the code here is useful enough to merge as-is, and I'd like to focus my time elsewhere for the time being if acceptable.

@AlexandreAbraham also mentioned adding some parameters to the `fetch_neurovault` function for simple filters. Pls let me know any that are desired, and I'll be glad to implement and test.

GaelVaroquaux commented:
> Currently, if a collection is updated in NeuroVault, it will not be updated
> locally. The only way to get around this is to pass the overwrite=True flag,
> which will re-download all metadata and images.

That's fine with me. It just needs to be documented.

> @AlexandreAbraham also mentioned adding some parameters to the
> fetch_neurovault function for simple filters. Pls let me know any that
> are desired,

I think it is. We care a lot about making simple things simple.

bcipolli commented:
>I think it is. We care a lot about making simple things simple.

Right :+1: Any suggestions for relevant filters? `collection_ids` (list of collections) stands out, perhaps `exclude_unpublished` (i.e. force a DOI)? Open to suggestions.

GaelVaroquaux commented:
> Right :+1: Any suggestions for relevant filters? collection_ids (list of
> collections) stands out, perhaps exclude_unpublished

Both of these, and maybe image_type.

bcipolli commented:
I added `collection_ids`, `image_ids`, `image_type`, `map_types`. `collection_ids` and  `image_ids` can be positive (include given IDs) or negative (exclude given IDs). 

I also migrated the examples to use these keyword args (where possible).

KamalakerDadi commented:
@bcipolli Your PR looks really good. Few comments from my side.

General comment on example: ```plot_ica_neurovault.py```
Example is really cool. But, for myself, it is a bit hard to follow. I am not sure so far if it is good to have lot of functions in a example. AFAIK some functions can be reduced by utilising functions from nilearn ? or Is it not ?

KamalakerDadi commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,260 @@
+
+
+@mem.cache
+def resample_and_expand_image(src_nii, target_nii, out_path=None, force=False):
```
```force``` is not used if you may have missed it.

KamalakerDadi replied:
Taking into account reviews on my PR's, I think ```src_nii``` should be denoted as ```src_img``` same to ```target_nii```.

KamalakerDadi commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,260 @@
+# Outline
+
+
+""" Code to grab the data from NeuroVault, and compute a map of
```
May be it is not necessary since you have already explained above ?

KamalakerDadi commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,260 @@
+"""
+NeuroVault cross-study ICA maps.
+=====================================================================
```
May be cutshorting "=".

KamalakerDadi commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,260 @@
+    data[np.isnan(data)] = 0
+    assert np.abs(data).max() > 0
+
+    # print('working on %s, %s' % (src_nii.get_filename(), len(np.unique(data))))
```
May be its not necessary ```# print(......```

KamalakerDadi commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,260 @@
+        if np.all([os.path.exists(p) for p in out_paths]):
+            return [nib.load(p) for p in out_paths]
+
+    # Copy the image and mkae a background mask.
```
typo "make"

lesteve commented:
It looks like this needs a rebase on master.

bcipolli commented:
Yep, a different version of the print/output changes were contained here. Just rebased to use the updated version of that code--thanks for merging!

bcipolli commented:
Rebased to master.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+
+
+
+# -------------------------------------------
```
The notebook-style examples have been merged.

Once rebased on master (which I think you have just done), and you turn all these '-' into '#', you should get pretty notebook-like rendering in the docs:
https://circle-artifacts.com/gh/nilearn/nilearn/757/artifacts/0/home/ubuntu/nilearn/doc/_build/html/auto_examples/connectivity/plot_adhd_spheres.html

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -1279,3 +1285,383 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    """
+
+    if url is None:
+        url = "http://neurovault.org/api"
```
I think that I would prefer this as a default in the function signature.

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -1279,3 +1285,383 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    if url is None:
+        url = "http://neurovault.org/api"
+    if collection_filters is None:
+        collection_filters = []
```
Could we have this as a default empty tuple (not list) in the function signature? To make things more explicit.

The good thing about a tuple is that it is not mutable.

bcipolli replied:
Actually, it's expected that `collection_filters` is mutable, as some default arguments get converted into `collection_filters`. Which means, if someone passes in a tuple, things will break.

I would prefer say `collection_filters = list(collection_filters)` to begin... is there a better way to handle this?

bcipolli replied:
Looks like `itertools.chain()` is the preferred method. I will convert these to tuples, in this case.

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -1279,3 +1285,383 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    if collection_filters is None:
+        collection_filters = []
+    if image_filters is None:
+        image_filters = []
```
Same remark about the tuple.

GaelVaroquaux commented on nilearn/datasets/utils.py:
```diff
@@ -532,15 +532,6 @@ def _fetch_file(url, data_dir, resume=True, overwrite=False,
-            e.reason = ("%s| Error while fetching file %s; "
-                          "dataset fetching aborted." % (
-                            str(e.reason), file_name))
-        raise
```
@AlexandreAbraham : can you comment on this change. This is code that you wrote originally, and you probably have a good feeling of where it is useful.

GaelVaroquaux commented:
Hey, @bcipolli , sorry for taking so long to look at this.

My main comment is that the examples are a bit scary. The fact that we come up with such examples suggests that we don't fully have the right abstractions. This is probably mostly unrelated with this PR (and something like https://github.com/nilearn/nilearn/issues/847 might help). However, the current examples are going against the will of nilearn to reassure users that things are simple and will be alright.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+mem = Memory('cache', verbose=0)
+
+
+@mem.cache
```
We don't need to cache this, do we?

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+def resample_and_expand_image(src_img, target_img, out_path=None):
+    """Resample src_img to target_img affine, turn 4D into list of 3D.
+
+    Optionally save to out_path.
```
I have a hunch that this function could be simplified using nilearn.image.iter_img. 

I am not sure quite how at a quick glance, but when I look at the function from a big picture, I have this feeling.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+            data, _ = _extrapolate_out_mask(data.astype(np.float), bg_mask,
+                                            iterations=3)
+    src_img = nib.Nifti1Image(data, src_img.get_affine(),
+                              header=src_img.get_header())
```
This could probably use new_img_like?

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+            # First save the files separately
+            this_nii = nib.Nifti1Image(resampled_data[..., index],
+                                       resampled_nii.get_affine())
+            resampled_niis.append(this_nii)
```
I believe that all the above (including the 'squeeze') should be possible to implement with iter_img. If it's not the case, we have gotten something wrong in the design of iter_img, and we need to fix that.

bcipolli replied:
One downside of `iter_img` is that it fails on a 3D image, rather than giving me an iterator for the one image. So I'll have to load the image and check if it's 4D before calling `iter_img`... that seems clunky to me.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+    # Save the result
+    if out_path:
+        for nii, nii_path in zip(resampled_niis, out_paths):
+            nib.save(nii, nii_path)
```
If we remove this file saving from the example, I am not sure that we loose a lot of functionality, and we do remove a lot of lines. What do you think?

bcipolli replied:
Agreed, good idea. With `mem.cache`, there's really no reason to additionally re-save the images, and most of the complexity of the function is in the save.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+    335,                           # 335 is a duplicate of 336
+    3360, 3362, 3364,              # These are mean images, and not Z maps
+    1202, 1163, 1931, 1101, 1099]  # Ugly / obviously not Z maps
+imfilts = [lambda im: im.get('perc_bad_voxels', 0.) < 100.]
```
Maybe we should special-case this filter in the fetch_neurovault function, if you think that it is very useful (it does seem so).

What do you think?

bcipolli replied:
Perhaps a `remove_known_bad_maps` flag that adds all of the filters above your comment.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+
+    Optionally save to out_path.
+    """
+    data = src_img.get_data().squeeze()
```
I think that if you used here 'check_niimg(..., ensure_ndim=3) you could pass in paths (which would simplify other parts of the example), and you would not need to squeeze.

It would indeed mean that you are importing from a private module, _'utils', which feels wrong, but we can expose a public path for this function if it is useful.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+        # If all output images exist, use them cached from disk.
+        if np.all([os.path.exists(p) for p in out_paths]):
+            return [nib.load(p) for p in out_paths]
+
```
Start comment (content of the comment below, sorry, I don't see a better way of doing this)

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+    src_img = nib.Nifti1Image(data, src_img.get_affine(),
+                              header=src_img.get_header())
+    del out_of_mask, bg_mask
+
```
End comment:

I think that all the lines between the start of the comment, and the end of the comment (and the line getting rid of nans that was before) could be put in a public function companion of _extrapolate_out_mask. Maybe 'extropolate_thresholded_img'. It feels like a function that would be of general use.

What do you think?

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+# -------------------------------------------
+# Get the grey matter mask. Cheat by pulling from github :)
+print("Downloading and resampling grey matter mask.")
+url = 'https://github.com/NeuroVault/neurovault_analysis/raw/master/gm_mask.nii.gz'
```
I think that this is a fairly universal brain mask that is obtained by thresholding the SPM tissue probability maps. We should either put it in the nilearn datasets, or (better) have downloaders for the SPM TPMs. @AlexandreAbraham : what do you think?

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,244 @@
+# -------------------------------------------
+# Do the ICA transform.
+masker = NiftiMasker(mask_img=mask, memory=mem)
+X = masker.fit_transform(resampled_niis)
```
I am wondering: maybe we can do the resampling using the NiftiMasker's resampling functionality? If that's possible, it would simplify the example.

bcipolli replied:
Oh, that's perfect--this makes memory caching easy too.

GaelVaroquaux commented:
I must now go to bed, and I won't be able to work on this before tomorrow afternoon. I have made many small comments on the example. I don't know if they make any sense. I was trying to figure out ways of simplifying it.

bcipolli commented:
Thanks @GaelVaroquaux ! It will be great to simplify the example and make good use of the `nilearn` tools. I've obviously been out of the `nilearn` loop too long, so my chops have diminished. Looking forward to beefing them up over the next couple of days.

Thanks for the detailed comments, and looking forward to getting pretty--and simple--examples done. :+1: 

P.S. @AlexandreAbraham congrats ;)

bcipolli commented:
@GaelVaroquaux I've updated the `plot_ica_neurovault` example to use more of the `nilearn` architecture, simplify the interface, and to push useful code into the fetcher function (useful filters, getting neurosynth terms). The example looks shorter and simpler to me now.

There are a few changes to core nilearn code that do need comments (including one small new one, to avoid a warning I was getting). I could pull those out and submit those separately, if it helps.

AlexandreAbraham commented:
We should probably not let this PR die. @bcipolli, do you want to finish it or should I take it over?

bcipolli commented:
@AlexandreAbraham I'm happy to push it forward. This needs to be rebased, there are three outstanding nilearn questions/changes that I'll summarize, and then there's the code/examples itself (which I think is in a reasonable state).

Will update once I've rebased.

bcipolli commented:
OK, rebased. Here's the situation on nilearn changes:

* [Add the http module](https://github.com/nilearn/nilearn/pull/832/files#diff-85f290884f840c5b3c6a87579852a5e9R18) to `_utils/compat`, as some http errors indicate offline status.
* [Add query_server arg to all functions](https://github.com/bcipolli/nilearn/commit/09b8e9371a8e785845c5f46212b8fed2640a7915) - `query_server` is important for neurovault. I pushed it to the `_fetch_files`. I also pushed it to each fetcher functions. I could easily pull this out as a separate PR... or scrap it, but need feedback.
* [Cleaning error verbosity](https://github.com/nilearn/nilearn/pull/832/files#diff-c22120ae66106d07e452aed0f968c08eL535) - this is a continuation of similar changes I made that have already been merged.
* [Change how `iter_image` handles 3D images](https://github.com/nilearn/nilearn/pull/832/files#diff-b61cedb53334db86de8881757209d3feL546) - failing on 3D images makes handling mixed datasets like neurovault more challenging. Is it important to fail, rather than do as requested?
* [Make sure to call _check_fitted()](https://github.com/nilearn/nilearn/pull/832/files#diff-f3684004932db308cd316be21301fb28R222) inside Masker functions.
* [Avoid needless warning](https://github.com/nilearn/nilearn/pull/832/files#diff-d906c149b6172542420c1cf50c27d14cL471) with a simple check.

Once those get reviewed specifically, the rest is about neurovault strictly, and so, easier. I will go check on test failures now.

AlexandreAbraham commented:
> Here's the situation on nilearn changes

I'm a bit afraid by the changes you propose here. I think we are going to run into the same problem as some of your other PR: the more you change, the longer it takes to merge it. For example, your PR could be stuck because you decided to change the behavior of `iter_img` (which is likely to happen). If you don't want your PR to be stuck in the process, please break it into smaller PRs (and maybe use a workaround if one of your proposition is waiting for validation).

> Add the http module

Can you be more explicit on the advantage of using it? Maybe add it as a comment where it's used?

> Add query_server arg to all functions

It may be useful for neurovault but I'm not sure that it is useful in all the other functions... If I understand what's happening, is one has downloaded a dataset from neurovault and that this dataset has changed post-download, `query_server=False` will make one uses the old atlas forever and `query_server=True` will check if a new version is available. This reminds me of `overwrite` a little bit. so +1 for neurovault, but +0 for the rest, I don't know if it can be useful.

>  Cleaning error verbosity

Can you say why you removed this piece of code?

>  Change how iter_image handles 3D images

Yes, this is a big deal in nilearn. We have taken special care to distinguish 3d from 4d data and all the uses of iter_img are based on the fact that it will raise an exception if 3D data is given. So modifying it as-is is not recommended. I am a bit puzzlled by the fact that you want to iterate over 3d image. Can you give an example?

>  Make sure to call _check_fitted()

Sure, but that should go in a separate PR.

> Avoid needless warning

Should we avoid a warning or raise an error?

GaelVaroquaux commented on examples/manipulating_visualizing/plot_neurovault_meta_analysis.py:
```diff
@@ -0,0 +1,99 @@
+import numpy as np
+from nilearn import datasets
+
+### Define filters ############################################################
```
Could you change these separators to use seperators like in the other examples of nilearn (one line of "#" and then the text below). That way sphinx-gallery knows how to split the example in a notebook-like example.

GaelVaroquaux commented on examples/manipulating_visualizing/plot_neurovault_meta_analysis.py:
```diff
@@ -0,0 +1,99 @@
+
+### Define filters ############################################################
+cfilts = []  # [lambda col: col['DOI'] is not None]
+imfilts = [lambda im: "stop signal" in (im.get('cognitive_paradigm_cogatlas') or "").lower(),
```
Using the ' or "" ', are you trying to use the pattern '.get(key, default)' ?

I would prefer to rely on that, because it is a standard Python pattern.

GaelVaroquaux replied:
Maybe it's also tell us that im (which is defined in the neurovault downloader) should be a defaultdict, with a default value of "".

GaelVaroquaux replied:
Scrap that. It would probably make debugging hard.

bcipolli replied:
Just for some quick feedback: this is the JSON dict from neurovault. It can return blank or `None`, or sometimes the value isn't present at all. The idiom above is how I typically handle such cases.

To simplify the example, I could try to normalize things inside the `fetch` function. Because I'm not sure what to expect from Neurovault (no documentation to that level) and after talking with @chrisfilo about the differences in metadata provided from different data sources, returning the raw JSON from `fetch` and letting users handle what they see, seemed like the safest thing.

GaelVaroquaux replied:
OK. I don't know what to conclude. I just find that the downloader is a
bit difficult to use because of these pecularities.

bcipolli replied:
The choices are: try to normalize, and risk that we'll get things wrong, be agnostic and let the data be peculiar to work with, or convince @chrisfilo to normalize things more easily across the board. I have a couple of issues open on it: https://github.com/NeuroVault/NeuroVault/issues/395 https://github.com/NeuroVault/NeuroVault/issues/396

I'm glad to go in whatever direction. I chose the direction that seemed seemed most flexible to me.

GaelVaroquaux replied:
> The choices are: try to normalize, and risk that we'll get things
> wrong, be agnostic and let the data be peculiar to work with, or
> convince @chrisfilo to normalize things more easily across the board. I
> have a couple of issues open on it: NeuroVault/NeuroVault#395
> NeuroVault/NeuroVault#396

The best option would be convincing @chrisfilo.

Please, pretty please with a cherry on top, @chrisfilo.

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -24,13 +33,22 @@ def fetch_haxby_simple(data_dir=None, url=None, resume=True, verbose=1):
+        Whether to resume download of a partly-downloaded file.
+
+    query_server: bool, optional (default: True)
+        if False, then only cached data is used.
```
I'd rather not add this to all the downloaders. It's going to lead to subtle bugs in what people are doing. I'd rather have the download just crash if there is no internet.

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -1406,3 +1507,521 @@ def fetch_megatrawls_netmats(dimensionality=100, timeseries='eigen_regression',
+            338, 339,                      # And the following are crap
+            335,                           # 335 is a duplicate of 336
+            3360, 3362, 3364,              # These are mean images, and not Z maps
+            1202, 1163, 1931, 1101, 1099]  # Ugly / obviously not Z maps
```
I think that "bad_collects" and 'bad_image_ids" should be module-level constants under the name "BAD_NEUROVAULT_COLLECTIONS and "BAD_NEUROVAULT_IMAGE_IDS"

GaelVaroquaux commented on nilearn/image/image.py:
```diff
@@ -536,14 +537,19 @@ def iter_img(imgs):
+        return check_niimg_4d(imgs, return_iterator=True)
+    except DimensionError:
+        # Wrap a 3D image in a dummy list.
+        return check_niimg_4d([imgs], return_iterator=True)
```
I am not sure how this is tied to your PR. Can you comment on where it is used?

bcipolli replied:
Indeed, I no longer use `iter_img` in `plot_ica_neurovault.py` after previous changes, so this is no longer needed. Will pull this and close the PR.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,121 @@
+print("Creating a grey matter mask.")
+target_img = datasets.load_mni152_template()
+grey_voxels = (target_img.get_data() > 0).astype(int)
+mask_img = new_img_like(target_img, grey_voxels)
```
We need a canonical brain mask added to nilearn. (issue created: #945)

KamalakerDadi replied:
FYI: mask is now available in nilearn, ```datasets.load_mni152_brain_mask()```

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,121 @@
+# Images may fail to be transformed, and are of different shapes,
+# so we need to trasnform one-by-one and keep track of failures.
+X = []
+xformable_idx = np.ones((len(images),), dtype=bool)
```
In terms of variable name, I'd prefer 'transformable'. It's unclear to a non native English speaker what this means.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,121 @@
+    """ Cast image to the specified dtype"""
+    img = check_niimg(img)
+    img_data = img.get_data().astype(dtype)
+    return new_img_like(img, img_data)
```
Maybe for the sake of simplicity you could merge clean_img and cast_img.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,121 @@
+    display.title(title, size=16)
+
+# Done.
+plt.show()
```
We prefer to use nilearn.plotting.show, because we are trying not to have to teach matplotlib to users if we don't have to.

GaelVaroquaux commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,121 @@
+    # Use the 4 terms weighted most as a title
+    important_terms = terms[np.argsort(ic_terms)[-4:]]
+    title = '%d: %s' % (idx, ', '.join(important_terms[::-1]))
+    display.title(title, size=16)
```
Is there a reason that you are not using the 'title' argument of plot_stat_map.

GaelVaroquaux replied:
The control of the size, maybe?

GaelVaroquaux commented on doc/whats_new.rst:
```diff
@@ -4,8 +4,11 @@
+- Small bugfix for more flexible input types (targetter in particular at
 making code easier in nistats).
+- New fetcher for NeuroVault data.
+  nilearn.datasets.fetch_neurovault
```
Maybe we should use the sphinx syntax ":func:`nilearn.datasets.neurovault`"

jeromedockes commented on examples/connectivity/plot_ica_neurovault.py:
```diff
@@ -0,0 +1,121 @@
+
+# Clean & report term scores
+terms = np.array(term_scores.keys())
+term_matrix = np.asarray(term_scores.values())
```
constructing arrays like this from dict.keys() and dict.values() doesn't seem to work for me in python 3, because these functions don't return lists anymore but dict_keys and dict_values objects. Maybe we should turn them into lists first, or access them in another way?

jeromedockes commented on nilearn/datasets/func.py:
```diff
@@ -1559,3 +1660,523 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+            1202, 1163, 1931, 1101, 1099]  # Ugly / obviously not Z maps
+        collection_ids = chain(collection_ids, [-bid for bid in bad_collects])
+        image_ids = chain(image_ids, [-bid for bid in bad_image_ids])
+        image_filters = chain(image_filters,
```
passing a dictionary as image_filters or collections_filters will not work because it is always used as if it were a tuple of callables (the other possibility for filtering) anyway. 
e.g.
`fetch_neurovault(max_images=10, collection_filters={'DOI': '10.1016/j.neurobiolaging.2012.11.002'})`

AlexandreAbraham commented:
Closed in favor of #1182.
